### ISSUE TITLE
[OP] Add cosine_similarity operation

### ISSUE THREAD

**Issue by benfred** [NVIDIA/NVTabular#39]
<a href="https://github.com/rnyak"><img src="https://avatars1.githubusercontent.com/u/16246900?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [rnyak](https://github.com/rnyak)**
_Tuesday May 19, 2020 at 22:15 GMT_
_Originally opened as https://github.com/rapidsai/recsys/issues/150_

----

**Is your operator request related to a problem? Please describe.**

Cosine similarity was used as a feature engineering technique in W&D Outbrain model. Cosine similarity is a metric used to measure similarity between two non-zero vectors of an inner product space.

**Describe the solution you'd like**

Apply cosine similarity as an operator.

A clear and concise description of the operation you'd like to perform on the column.  Please include:
 - Type: Feature Engineering 
 - input column type(s): Continuous numeric (X and Y as two vectors)
 - output column type(s): Continuous numeric within [-1, 1]
 - Expected transformation of the data after application: The resulting similarity ranges from −1 meaning exactly opposite, to 1 meaning exactly the same.

**Optional: Describe operation stages in detail***
Apply: `compute cosine similarity of v1 to v2: (v1 dot v2)/{||v1||*||v2||)` see [wikipedia](https://en.wikipedia.org/wiki/Cosine_similarity) for more information.

**Additional context**
Sklearn has a cosine_similarity function:

`sklearn.metrics.pairwise.cosine_similarity(X, Y=None, dense_output=True)` [source](https://github.com/scikit-learn/scikit-learn/blob/fd237278e/sklearn/metrics/pairwise.py#L1143).

One alternative way is to calculate `cosine_distance` and then apply `1- cosine_distance` to calculate `cosine_similarity` value.

Currently `cuml` does not have python api or c++ api for `cosine distance`. There is only the prims API. The prims API are header only.  The CUDA headers can be found in the following link:

https://github.com/rapidsai/cuml/tree/4084790afe605e82710597be575b63d8f57b1bbb/cpp/src_prims/distance

Wrappers need to be crated both in libcuml for easy C++ consumption as well as python.

**Comment by benfred** [comment]
<a href="https://github.com/alecgunny"><img src="https://avatars0.githubusercontent.com/u/14932242?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [alecgunny](https://github.com/alecgunny)**
_Wednesday May 20, 2020 at 16:46 GMT_

----

How would this functionality look in the context of tabular data without array-element support? Would you just specify two groups of columns, the elements within each group specifying elements of the vectors?

**Comment by gabrielspmoreira** [comment]
A common usage of cosine similarity in the context of RecSys, is to pick a list of users (whose profile is represented by vectors) and computing the pairwise similarity between those users and all recommendable items.
In that case, it is usually more efficient to implement cosine similarity by applying l2 normalization (like sklearn.preprocessing.normalize(X, norm='l2') on all vectors first (just once), and then compute the dot product between the users matrix and the items matrix.

**Comment by kkranen** [comment]
Here's an example of the minimal representation of my implementation of tablewise cosine similarity:
`d = {'group1': [1, 2], 'group2': [3, 4]}
train_sample = cudf.DataFrame(data=d)

d = {'groups': [1, 3,1,2,1,2,3], 'ids':[3,3,8,9,4,3,6],'values': [.1, .4,.1, .4,.1, .4,.1]}
categories_sample = cudf.DataFrame(data=d)

def cosine_similarity(merged_gdf,x_gdf, y_gdf, index_x, index_y,value_x ,value_y,x_join, merge_x_join, y_join, merge_y_join):
    x_norm = x_gdf
    x_norm['squared_value_x'] = x_norm[value_x]**2

    x_norm= x_norm.groupby([x_join])[['squared_value_x']].agg('sum')
    x_norm['squared_value_x'] = np.sqrt(x_norm['squared_value_x'])
    x_norm.reset_index(inplace=True)
    
    y_norm = y_gdf
    y_norm['squared_value_y'] = y_norm[value_y]**2
    y_norm= y_norm.groupby(y_join)[['squared_value_y']].agg('sum')
    y_norm['squared_value_y'] = np.sqrt(y_norm['squared_value_y'])
    y_norm.reset_index(inplace=True)
    
    
    merged_gdf = merged_gdf[[merge_x_join,merge_y_join]].drop_duplicates().merge(x_gdf,left_on=merge_x_join,right_on=x_join). \
        merge(y_gdf,left_on=merge_y_join,right_on=y_join)
    
    if index_x == index_y:
        index_x += '_x'
        index_y += '_y'
    if value_x == value_y:
        value_x += '_x'
        value_y += '_y'
    merged_gdf['product'] = merged_gdf[value_x] * merged_gdf[value_y]
    merged_gdf.loc[merged_gdf[index_x] != merged_gdf[index_y],'product'] = 0
    merged_gdf = merged_gdf.groupby([merge_x_join,merge_y_join])[['product']].agg('sum')
    merged_gdf.reset_index(inplace=True)
    
    
    
    merged_gdf = merged_gdf.merge(x_norm, left_on= merge_x_join, right_on=x_join)\
        .merge(y_norm, left_on=merge_y_join,right_on=y_join)
    
    merged_gdf['cosine_similarity'] = merged_gdf['product'] / (merged_gdf['squared_value_x'] * merged_gdf['squared_value_y'])
    return merged_gdf

cosine_similarity(train_sample,categories_sample, categories_sample, 'ids', 'ids','values' ,'values','groups', 'group1', 'groups', 'group2')

`
